docs/dxp/7.x/en/installation-and-upgrades/setting-up-liferay-dxp/tuning-your-jvm.md

```diff
@@ -0,0 +1,103 @@
+# Tuning Your JVM
+
+Java Virtual Machine (JVM) tuning primarily focuses on adjusting the garbage collector and the Java memory heap. We used Oracle's 1.8 JVM for the reference architecture. You may choose other supported JVM versions and implementations. Please consult the  [Liferay DXP Compatibility Matrix](https://web.liferay.com/group/customer/dxp/support/compatibility-matrix) for additional compatible JVMs.
+
+Here are the JVM tuning topics:
+
+* [Garbage Collector](#garbage-collector)
+* [Code Cache](#code-cache)
+* [Java Heap](#java-heap)
+* [JVM Advanced Options](#jvm-advanced-options)
+
+Garbage collection is first.
+
+## Garbage Collector
+
+Choosing the appropriate garbage collector (GC) helps improve the responsiveness of your Liferay DXP deployment. Use the concurrent low pause collectors:
+
+```bash
+-XX:+UseParNewGC -XX:ParallelGCThreads=16 -XX:+UseConcMarkSweepGC
+-XX:+CMSParallelRemarkEnabled -XX:+CMSCompactWhenClearAllSoftRefs
+-XX:CMSInitiatingOccupancyFraction=85 -XX:+CMSScavengeBeforeRemark
+```
+
+You may choose from other available GC algorithms, including parallel throughput collectors and G1 collectors. Start tuning using parallel collectors in the new generation and concurrent mark sweep (CMS) in the old generation.
+
+**Note:** the `ParallelGCThreads` value (e.g., `ParallelGCThreads=16`) varies  based on the type of CPUs available. Set the value according to CPU specification. On Linux machines, report the number of available CPUs by running `cat /proc/cpuinfo`.
+
+**Note:** There are additional "new" algorithms like G1, but Liferay Engineering's tests for G1 indicated that it does not improve performance. Since your application performance may vary, you should add G1 to your testing and tuning plans.
+
+## Code Cache
+
+Java's just-in-time (JIT) compiler generates native code to improve performance. The default size is `48m`. This may not be sufficient for larger applications. Too small a code cache reduces performance, as the JIT isn't able to optimize high frequency methods. For Liferay DXP,  start with `64m` for the initial code cache size.
+
+```bash
+-XX:InitialCodeCacheSize=64m -XX:ReservedCodeCacheSize=96m
+```
+
+Examine the efficacy of the parameter changes by adding the following logging parameters:
+
+```bash
+-XX:+PrintCodeCache -XX:+PrintCodeCacheOnCompilation
+```
+
+## Java Heap
+
+When most people think about tuning the Java memory heap, they think of setting the maximum and minimum memory of the heap. Unfortunately, most deployments require far more sophisticated heap tuning to obtain optimal performance, including tuning the young generation size, tenuring durations, survivor spaces, and many other JVM internals.
+
+For most systems, it's best to start with at least the following memory settings:
+
+```bash
+server -XX:NewSize=700m -XX:MaxNewSize=700m -Xms2560m -Xmx2560m -XX:MetaspaceSize=512m
```

Comment:
@jhinkey Current master head: `-Xms2560m -Xmx2560m -XX:MaxNewSize=1536m -XX:MaxMetaspaceSize=768m -XX:MetaspaceSize=768m -XX:NewSize=1536m`

Reply:
@jhinkey 7.3.x head is also like what I commented. See https://github.com/liferay/liferay-portal-ee/blob/7.3.x/tools/servers/tomcat/bin/setenv.sh

Reply:
@dantewang What are the hardware specs (CPUs, RAM, ???) relevant to this tomcat configuration?

Reply:
@jhinkey This is the default value of portal, so anyone who downloads portal Tomcat bundle (CE or DXP) will have these by default. It's a starting point for most developer machines.

For production, it varies a lot based on not only their hardware specs, but also their use case. For Liferay benchmark tests, we use `-Xmx18g -Xms18g -XX:NewSize=9g -XX:MaxNewSize=9g -XX:MetaspaceSize=1g -XX:MaxMetaspaceSize=1g`. We found that setting new size to half of the total heap provides better performance than using a smaller new size. When using these settings, we also make sure the total large page size (`cat /proc/meminfo` -> HugePages_Total * Hugepagesize) can contain all of JVM's memory usage.

Reply:
@jhinkey The benchmark memory layout I mentioned above is used in different tiers of machines, including racked servers with 12C24T CPU & 64GB ECC memory (32G large page) and desktop-level machines with 4C8T CPU and 32GB memory (20G large page I think).